### PR TITLE
Implement `FileProxy` and `FileAttachment` objects

### PR DESCRIPTION
**Pull request by jathanism**
### Fixes: #n/a

These are implemented for the purpose of storing files in the database for use with `Jobs` and `FileVar` fields. The `FileVar` fields will now serialize files into the database store the PK of the file in the job metadata for pickup and retrieval of file contents by the worker.

- `django-db-file-storage` added as a dependency for use with `FileProxy` objects
- `FileProxy` is a new model that holds the filename and reference to the file contents and store them in the database using the `DatabaseFileStorage` backend provided by `django-db-file-storage`
- `FileAttachment` is a new model that holds the file contents and metadata and represents files on the datatbase-backed virtual filesystem. It is not intended for direct usage.
- Model tests & functional unit tests
- Modified `nautobot.utilities.utils.copy_safe_request` to no longer pass `request.FILES` through as files cannot be reliably serialized and we solved file serialization at the `BaseJob` class
- Made the following changs to `BaseJob`:
  - New `save_file()` to save a file handle to a `FileProxy` objects and return the UUID pk
  - New `load_file()` to load a `FileProxy` object by UUID and return a file handlek
  - New `_get_file_vars()` to return a mapping of all job fields that are of type `FileVar`
  - New `delete_files()` to delete `FileProxy` objects by a list of UUIDs
- The `run_job` function now does the following:
  - Uses `job._get_file_vars()` to generate a list of UUIDs to use for `FileProxy` cleanup
  - `FileProxy` objects are deleted on any failure scenario, or on successful job completion
  - Replaced processing `request.FILES` with `job.deserialize_data()/serialize_data()`
- Made the following changes to `FileProxy` objects:
  - Added a `/files/` root URL for downloading of files
  - Added an admin dashboard page until we come up with something better
  - Added `FileProxy.Meta.verbose_name_plural` as "file proxies"
- Added `JobResult.Meta.get_latest_by = "created"` so we can take advantage of queryset `earliest()/latest()` methods
- Added `CELERY_RESULT_ACCEPT_CONTENT = ["nautobot_json"]` to core settings to explicitly assert only `nautobot_json` as a supported task result serializer

### CODE REVIEW

File: nautobot/extras/tests/test_models.py [hunk 1]
```diff
@@ -726,3 +729,30 @@ def test_render_method(self):
     def test_render_method_bad_template(self):
         rendered_value = self.bad_computed_field.render(context={"obj": self.site1})
         self.assertEqual(rendered_value, self.bad_computed_field.fallback_value)
+
+
+class FileProxtyTest(TestCase):
```
Comment (jathanism) [inline nautobot/extras/tests/test_models.py:734]: PROXTY

File: nautobot/extras/jobs.py [hunk 2]
```diff
@@ -296,17 +298,53 @@ def serialize_data(data):
 
         These are converted back during job execution.
         """
+
         return_data = {}
         for field_name, value in data.items():
+            # MultiObjectVar
             if isinstance(value, QuerySet):
                 return_data[field_name] = list(value.values_list("pk", flat=True))
+            # Objectvar
```
Comment (glennmatthews) [inline nautobot/extras/jobs.py:307]:
```suggestion
            # ObjectVar
```

File: nautobot/extras/models/__init__.py [hunk 3]
```diff
@@ -23,20 +25,22 @@
...
+    "FileProxy",
     "GitRepository",
     "GraphQLQuery",
     "ImageAttachment",
     "Job",
     "JobResult",
     "ObjectChange",
+    "Status",
+    "StatusField",
+    "StatusModel",
```
Comment (glennmatthews) [inline nautobot/extras/models/__init__.py:43]: `Status*` probably belongs after `Relationship*`, as long as you're moving it around anyway. :-)

**Review by glennmatthews — COMMENTED** [review]
Would it be useful to provide a basic set of views (even just using `admin` for now, though I know we're trying to move away from it) for listing/inspecting/managing FileAttachment and FileProxy records? Or are we okay for now just using nbshell to inspect and clean up as needed?

**Comment by jathanism** [comment]
> Would it be useful to provide a basic set of views (even just using `admin` for now, though I know we're trying to move away from it) for listing/inspecting/managing FileAttachment and FileProxy records? Or are we okay for now just using nbshell to inspect and clean up as needed?

I think for now having the admin is suitable until we decide what to do.

File: nautobot/extras/tests/test_jobs.py [hunk 4]
```diff
@@ -176,3 +176,22 @@ def test_read_only_no_commit_field(self):
 <input class="form-control form-control" id="id_var" name="var" placeholder="None" required type="text">
 <br><span class="helptext">Hello</span><input id="id__commit" name="_commit" type="hidden" value="False"></td></tr>""",
             )
+
+
+class JobFileUploadTest(TestCase):
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_jobs.py:181]: Possibly need to add

```python
@skipIf(
    "dummy_plugin" not in settings.PLUGINS,
    "dummy_plugin not in settings.PLUGINS",
)
```

?

Comment (jathanism) [reply nautobot/extras/tests/test_jobs.py:181]: Oh for sure. Thank you. I really wanted to knock this out last night but I hit a wall!

File: nautobot/extras/models/models.py [hunk 5]
```diff
@@ -317,10 +319,70 @@ def clean(self):
...
+class FileProxy(BaseModel):
+    """An object to store a file in the database.
+
+    The `file` field can be used like a file handle. The file contents are stored and retrieved from
+    `FileAttachment` objects.
+
+    The associated `FileAttachment` is removed when `delete()` is called. For this reason, one
+    should never use bulk delete operations on `FileProxy` objects, unless `FileAttachment` objects
+    are also bulk-deleted, because a model's `delete()` method is not called during bulk operations.
+    In most cases, it is better ot iterate over a queryset of `FileProxy` objects and call
```
Comment (jathanism) [inline nautobot/extras/models/models.py:358]:
```suggestion
    In most cases, it is better to iterate over a queryset of `FileProxy` objects and call
```

File: examples/dummy_plugin/dummy_plugin/jobs.py [hunk 6]
```diff
@@ -1,4 +1,4 @@
-from nautobot.extras.jobs import Job
+from nautobot.extras.jobs import FileVar, Job, StringVar
```
Comment (jathanism) [inline examples/dummy_plugin/dummy_plugin/jobs.py:1]:
```suggestion
from nautobot.extras.jobs import Job
```

File: nautobot/extras/tests/test_jobs.py [hunk 7]
```diff
@@ -176,3 +178,88 @@ def test_read_only_no_commit_field(self):
...
+            )
+
+            # Serialize the file to FileProxy
+            data = {"file": self.dummy_file}
+            serialized_data = job.serialize_data(data)
+
+            # Assert that the file was serialized to a FileProxy
+            self.assertTrue(isinstance(serialized_data["file"], uuid.UUID))
+            self.assertTrue(serialized_data["file"], FileProxy.objects.latest().pk)
+            self.assertTrue(FileProxy.objects.count(), 1)
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_jobs.py:216]:
```suggestion
            self.assertEqual(serialized_data["file"], FileProxy.objects.latest().pk)
            self.assertEqual(FileProxy.objects.count(), 1)
```

File: nautobot/extras/tests/test_jobs.py [hunk 8]
```diff
@@ -176,3 +178,88 @@ def test_read_only_no_commit_field(self):
...
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_jobs.py:251]:
```suggestion
            self.assertEqual(serialized_data["file"], FileProxy.objects.latest().pk)
            self.assertEqual(FileProxy.objects.count(), 1)
```

File: nautobot/extras/tests/test_jobs.py [hunk 9]
```diff
@@ -1,13 +1,15 @@
 import os
 import uuid
 
+from django.core.files.uploadedfile import SimpleUploadedFile
 from django.conf import settings
 from django.contrib.contenttypes.models import ContentType
+from django.test import override_settings
```
Comment (glennmatthews) [inline nautobot/extras/tests/test_jobs.py:7]: Unused import?

File: nautobot/core/urls.py [hunk 10]
```diff
@@ -45,6 +45,8 @@
     path("", include("social_django.urls", namespace="social")),
     # django-health-check
     path(r"health/", include("health_check.urls")),
+    # FileProxy attachments
+    path("files/", include("db_file_storage.urls")),
```
Comment (glennmatthews) [inline nautobot/core/urls.py:49]: Not being familiar with db_file_storage, what views does this add? Are they useful, do we need to document them somewhere and/or add them to a nav menu somewhere?

File: nautobot/extras/tests/test_jobs.py (hunk 7 above)
Comment (jathanism) [reply nautobot/extras/tests/test_jobs.py:216]: Oh whoops.

File: nautobot/core/urls.py (hunk 10 above)
Comment (jathanism) [reply nautobot/core/urls.py:49]: Just basic views URLS for retrieving/downloading the files. I had to add them for the admin integration to be able to have detail views for the files.

**Review by glennmatthews — APPROVED** [review]

**Pull request merged**